### PR TITLE
Some timeout refactoring

### PR DESCRIPTION
**Pull request by bobcatfish**
# Changes

## Add more details about how the timeout handling works 🕒

While investigating #2905, I struggled to understand how the timeout
handling works, especially with TimeoutSet having very little comments,
so I've added some. I didn't add anything for backoffs yet because I'm
hoping we can separate that into a separate structure since it has a
very specific purpose that doesn't generalize to all timeouts.

Also changed the name "finished" to consistently use "done" so the
reader doesn't have to wonder about the difference between "finished"
and "done" (there isn't one)

## Move timeout handler into its own package 📦

I'd like to move the "backoff" logic into its own file, separate from
the other timeout logic, so it's clear which parts apply to what (i.e.
the timeout handler is being used for 2 purposes: timing out Runs which
take too long, and backing off when pod creation is failing - this is
totally fine but it's hard to understand when reading the code)

As a first step, I've moved the timeout handler into a separate package,
so we can have a file and tests dedicated to the backoff logic separate
from the other handling.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

 ```release-note
 NONE
 ```

### CODE REVIEW

File: pkg/reconciler/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ var (
 	_ pipelinerunreconciler.Interface = (*Reconciler)(nil)
 )
 
-// Reconcile compares the actual state with the desired, and attempts to
+// ReconcileKind compares the actual state with the desired, and attempts to
```
Comment (dlorenc) [inline pkg/reconciler/pipelinerun/pipelinerun.go:127]: Nice catch :)

File: pkg/timeout/handler.go [hunk 2]
```diff
@@ -58,24 +58,37 @@ type Backoff struct {
...
+// Handler knows how to track channels that can be used to communicate with go routines
+// which timeout, and the functions to call when that happens.
+type Handler struct {
+	logger *zap.SugaredLogger
+
+	// taskRunCallbackFunc is the function to call when a TaskRun has timed out.
+	// This is usually set to the function that enqueues the taskRun for reconciling.
+	taskRunCallbackFunc func(interface{})
+	// pipelineRunCallbackFunc is the function to call when a TaskRun has timed out
+	// This is usually set to the function that enqueues the taskRun for reconciling.
```
Comment (pritidesai) [inline pkg/timeout/handler.go:70]: NIT: 
s/when a TaskRun/when a PipelineRun/
s/enqueues the taskRun/enqueues the pipelineRun/

File: pkg/timeout/handler.go [hunk 3]
```diff
@@ -274,7 +286,7 @@ func (t *TimeoutSet) waitRun(runObj StatusKey, timeout time.Duration, startTime
 // the lifetime of the TaskRun no resources are released after the timer
 // fires. It is the caller's responsibility to Release() the TaskRun when
 // work with it has completed.
-func (t *TimeoutSet) SetTaskRunTimer(tr *v1beta1.TaskRun, d time.Duration) {
+func (t *Handler) SetTaskRunTimer(tr *v1beta1.TaskRun, d time.Duration) {
```
Comment (pritidesai) [inline pkg/timeout/handler.go:289]: Random thought, `SetTaskRunTimer` but no `SetPipelineRunTimer` 🤔

File: pkg/timeout/handler.go [hunk 4]
```diff
@@ -186,7 +198,7 @@ func (t *TimeoutSet) checkPipelineRunTimeouts(namespace string, pipelineclientse
 
 // CheckTimeouts function iterates through a given namespace or all namespaces
 // (if empty string) and calls corresponding taskrun/pipelinerun timeout functions
-func (t *TimeoutSet) CheckTimeouts(namespace string, kubeclientset kubernetes.Interface, pipelineclientset clientset.Interface) {
+func (t *Handler) CheckTimeouts(namespace string, kubeclientset kubernetes.Interface, pipelineclientset clientset.Interface) {
```
Comment (pritidesai) [inline pkg/timeout/handler.go:201]: its interesting that we are scrapping through all possible namespaces (if not specified) and checking all `TaskRuns` and `PipelineRuns` in those namespaces or at least in one specified namespace 😲

**Comment by pritidesai** [comment]
@bobcatfish I learnt some timeout handler with this PR 😜  and excited to see more changes ...

one minor [NIT](https://github.com/tektoncd/pipeline/pull/3011#discussion_r460334610) which can be addressed with next set of changes

/lgtm

File: pkg/timeout/handler.go (hunk 4 above)
Comment (pritidesai) [reply pkg/timeout/handler.go:201]: And the same check `CheckTimeouts` is done in TaskRun controller and PipelineRun controller 🤔

File: pkg/timeout/handler.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/timeout/handler.go:70]: whoops, thanks @pritidesai !

File: pkg/timeout/handler.go (hunk 4 above)
Comment (bobcatfish) [reply pkg/timeout/handler.go:201]: yeah!! the second means we're probably doing this twice as much as we need to 😅 

and it sounds like we probably don't need to be doing it at all!! https://github.com/tektoncd/pipeline/issues/2905#issuecomment-663717912

**Pull request merged**